block-storage/src/main/scala/coop/rchain/blockstorage/dag/merging/ConflictResolutionLogic.scala

```diff
@@ -0,0 +1,149 @@
+import coop.rchain.blockstorage.dag.merging.data._
+import coop.rchain.shared.syntax._
+
+import scala.collection.compat.immutable.LazyList
+
+object ConflictResolutionLogic {
+
+  // Remove from target deploy chains that depend on dependencies.
+  def removeDependentChains[D](
+      target: Set[D],
+      dependencies: Set[D]
+  )(depends: (D, D) => Boolean): Set[D] = {
+    type Result = (Set[D], Set[D], Set[D])
+    val init = (target, dependencies, Set.empty[D])
+    val (_, _, totalRejected) = LazyList
+      .unfold[Result, Result](init) {
+        case (acc, toReject, rejectedAcc) =>
+          val depending = acc.filter(d => toReject.exists(depends(d, _)))
+          val rOpt = if (depending.nonEmpty) {
+            val newAcc         = acc -- depending
+            val newRejectedAcc = rejectedAcc ++ depending
+            (newAcc, depending, newRejectedAcc).some
+          } else none[Result]
+          rOpt.map(r => (r, r))
+      }
+      .lastOption
+      .getOrElse(init)
+    totalRejected
+  }
+
+  // Remove incompatible deploys
+  // - deploys that conflicting with not seen but final
+  // - deploys that depend on finally rejected
+  // withDependencies function serves to clear dependencies of incompatible deploys within the conflict set
+  def enforceFinality[D](conflictSet: Set[D])(
+      acceptedFinally: Set[D],
+      rejectedFinally: Set[D],
+      conflicts: (D, D) => Boolean,
+      depends: (D, D) => Boolean,
+      withDependencies: Set[D] => Set[D]
+  ): Set[D] = {
+    val incompatible = conflictSet.filter { m =>
+      rejectedFinally.exists(depends(m, _)) || acceptedFinally.exists(conflicts(m, _))
+    }
+    withDependencies(incompatible)
+  }
+
+  /** Compute enforced rejections and choose [[MergingStrategy]] */
+  def computeMergingStrategy[B: Ordering, D](
+      latestMsgs: Set[B],
+      conflictScope: Set[B],
+      enforceReject: Set[D]
+  )(seen: B => Set[B], deploys: B => Set[D]): MergingStrategy[B, D] = {
+    // compute tip views on a conflict set given rejection enforcement
+    val tipViews = latestMsgs.map { m =>
+      val mergeTarget = (seen(m) intersect conflictScope).flatMap(deploys)
+      val toMerge     = mergeTarget -- enforceReject
+      val altered     = mergeTarget != toMerge
+      m -> (toMerge, altered)
+    }
+    // find tips which state that does not contain deploys from enforcedRejections
+    val tipsNotAltered = tipViews.collect { case (id, (_, altered)) if !altered => id }
+    // if at least one tip state is not altered - we can merge into this tip
+    if (tipsNotAltered.nonEmpty) {
+      val toMerge = tipViews.map { case (id, (view, _)) => (id, view) }
+      MergeIntoTip(toMerge, tipsNotAltered, enforceReject)
+    } // otherwise only merge into final fringe is possible, no previous conflict resolution can be reused
+    else {
+      val toMerge = tipViews.flatMap { case (_, (toMerge, _)) => toMerge }
+      MergeIntoFringe(toMerge, enforceReject)
```

Comment:
Why this cannot be only one type and in this case `tipsNotAltered` would be empty?

Reply:
toMerge here is just a plain Set - the whole conflict set, in case of MergeIntoTips toMerge is Set[(B, Set[D]) - views of each tip on a conflict set.